File: 494. Target Sum-2.cpp
class Solution {
public:
  int findTargetSumWays(vector<int> &nums, int target) {
    unordered_map<int, int> umap, temp; // i, count
    umap[0] = 1;

    for (auto x : nums) {
      temp.clear();
      for (auto [i, count] : umap) {
        temp[i + x] += count;
        temp[i - x] += count;
      }
      swap(umap, temp);
    }

    return umap[target];
  }
};

File: 494. Target Sum-2.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        # p
        # s - p
        # p - (s - p) = t
        # 2p = s + t
        # p = (s + t) / 2
        target += sum(nums)
        if target < 0 or target % 2:
            return 0
        target //= 2

        n = len(nums)

        f = [[0] * (target + 1) for _ in range(n + 1)]
        f[0][0] = 1

        for i, x in enumerate(nums):
            for c in range(target + 1):
                if c < x:
                    f[i + 1][c] = f[i][c]
                else:
                    f[i + 1][c] = f[i][c] + f[i][c - x]

        return f[n][target]
